File: src/main/java/com/joraph/loader/EntityLoaderDescriptor.java
package com.joraph.loader;

public class EntityLoaderDescriptor<A, I, R> {

	private final Class<R> entityClass;
	private final Class<A> argumentClass;
	private final LoaderFunction<A, I, R> loader;

	public EntityLoaderDescriptor(
			Class<R> entityClass,
			Class<A> argumentClass,
			LoaderFunction<A, I, R> loader) {
		this.entityClass = entityClass;
		this.argumentClass = argumentClass;
		this.loader = loader;
	}

	public Class<R> getEntityClass() {
		return entityClass;
	}

	public Class<A> getArgumentClass() {
		return argumentClass;
	}

	public LoaderFunction<A, I, R> getLoader() {
		return loader;
	}

	public boolean requiresAdditionalArguments() {
		return argumentClass != null;
	}

}

File: src/main/kotlin/com/joraph/loader/EntityLoaderDescriptor.kt
package com.joraph.loader

class EntityLoaderDescriptor<ArgProvider, Arg, ID, Entity>(
    val argumentProviderClass: Class<ArgProvider>? = null,
    val argumentExtractor: ArgumentExtractor<ArgProvider, Arg>? = null,
    val entityClass: Class<Entity>,
    val loader: LoaderFunction<Arg, ID, Entity>) {

    fun requiresAdditionalArguments(): Boolean {
        return argumentProviderClass != null && argumentExtractor != null
    }
}
